### PR TITLE
CSP OAuth Token Support

### PR DESCRIPTION
**Pull request by avanbrunt-cb**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Tests have been added that prove the fix is effective or that the feature works.
- [x] New and existing tests pass locally with the changes.
- [x] Code follows the style guidelines of this project (PEP8, clean code, linter).
- [x] A self-review of the code has been done.

## What is the ticket or issue number?

CBAPI-2553

## Pull Request Description

Add CSP OAuth tokens to Credential

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## How Has This Been Tested?

Manual integration tests and existing unit tests

### CODE REVIEW

**Review by lhilmes-cb — APPROVED** [review]
Can't really confirm if everything is right, but here's an approve if you need it!

File: src/cbc_sdk/credentials.py [hunk 1]
```diff
@@ -158,6 +179,66 @@ def to_dict(self):
...
+            OAUTH_APP_TOKEN_URL = f"{CSP_URL}/csp/gateway/am/api/auth/token"
+
+            # Construct Authorization header Basic Base64(client_id:client_secret)
+            client_id = self._values[CredentialValue.CSP_OAUTH_APP_ID]
+            client_secret = self._values[CredentialValue.CSP_OAUTH_APP_SECRET]
+            client_credentials = base64.b64encode(bytes(f"{client_id}:{client_secret}", "utf-8"))
+            headers = {"Authorization": f"Basic {client_credentials.decode('utf-8')}"}
+
+            resp = requests.post(OAUTH_APP_TOKEN_URL, {"grant_type": "client_credentials"}, headers=headers)
+            json_body = resp.json()
```
Comment (emitreva-cb) [inline src/cbc_sdk/credentials.py:231]: I am not 100% sure, but is it possible that upon certain errors this will raise an exception?

Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py:231]: I am not aware of non-json errors at the moment but I can add a try except though I am not sure which exception I would want to raise instead

Comment (emitreva-cb) [reply src/cbc_sdk/credentials.py:231]: I was kind of thinking more like moving this line: `json_body = resp.json()` after the check for != 200

Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py:231]: I can't because I included the error message in the CredentialError Exception

**Review by rmcqueen-cb — COMMENTED** [review]
The logic, implementation, and tests all look solid. I've mentioned a few enhancements for conventionality.

File: src/cbc_sdk/credentials.py [hunk 2]
```diff
@@ -158,6 +179,66 @@ def to_dict(self):
...
+        elif self._token_type == "API_KEY":
+            return self._values[CredentialValue.TOKEN]
+
+        CSP_URL = self._values.get(CredentialValue.CSP_URL_OVERRIDE).rstrip("/")
+
+        if self._token_type == "API_TOKEN":
+            API_TOKEN_URL = f"{CSP_URL}/csp/gateway/am/api/auth/api-tokens/authorize"
+            resp = requests.post(API_TOKEN_URL, {"api_token": self._values[CredentialValue.CSP_API_TOKEN]})
+            json_body = resp.json()
+            if resp.status_code != 200:
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py:213]: I aired on the side of specificity given that I only expect 200 responses if I ever receive a non 200 response less than 400 then something is going to break inside the if statement

File: src/cbc_sdk/credentials.py [hunk 3]
```diff
@@ -158,6 +179,66 @@ def to_dict(self):
...
+        if self._token_type == "OAUTH_APP":
+            OAUTH_APP_TOKEN_URL = f"{CSP_URL}/csp/gateway/am/api/auth/token"
+
+            # Construct Authorization header Basic Base64(client_id:client_secret)
+            client_id = self._values[CredentialValue.CSP_OAUTH_APP_ID]
+            client_secret = self._values[CredentialValue.CSP_OAUTH_APP_SECRET]
+            client_credentials = base64.b64encode(bytes(f"{client_id}:{client_secret}", "utf-8"))
+            headers = {"Authorization": f"Basic {client_credentials.decode('utf-8')}"}
+
+            resp = requests.post(OAUTH_APP_TOKEN_URL, {"grant_type": "client_credentials"}, headers=headers)
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py:230]: Good to know. It's always interesting to see the shorthand they add into the library

File: src/tests/uat/csp_oauth.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+If you don't have CSP credentials follow this [guide](TBD)
+"""
+
+import sys
+
+from cbc_sdk import CBCloudAPI
+from cbc_sdk.platform import Device
+
+url = ''
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/tests/uat/csp_oauth.py]: I've done the first part however the second would be dependent on that token having device permissions which may not always be the case

File: src/cbc_sdk/credentials.py [hunk 5]
```diff
@@ -158,6 +179,66 @@ def to_dict(self):
...
+            if resp.status_code != 200:
+                raise CredentialError(json_body.get("message"))
+
+            # Check that token has scope
+            if json_body.get("scope") is None:
+                raise CredentialError("Access token created with empty scope")
+            return json_body.get("access_token")
+
+        if self._token_type == "OAUTH_APP":
+            OAUTH_APP_TOKEN_URL = f"{CSP_URL}/csp/gateway/am/api/auth/token"
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py]: Done

File: src/cbc_sdk/credentials.py [hunk 6]
```diff
@@ -158,6 +179,66 @@ def to_dict(self):
...
+        """
+        Get token required to authenticate with VMware Carbon Black Cloud
+
+        Returns:
+            str: Token string for VMware Carbon Black Cloud
+        """
+        if self._token_type == "UNKNOWN":
+            return None
+        elif self._token_type == "API_KEY":
+            return self._values[CredentialValue.TOKEN]
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py]: Done

File: src/cbc_sdk/credentials.py [hunk 7]
```diff
@@ -158,6 +179,66 @@ def to_dict(self):
...
+
+        CSP_URL = self._values.get(CredentialValue.CSP_URL_OVERRIDE).rstrip("/")
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py]: Done

File: src/cbc_sdk/connection.py [hunk 8]
```diff
@@ -177,10 +177,13 @@ def __init__(self,
             ApiError: If there's an internal error initializing the connection.
             ConnectionError: If there's a problem with the credentials.
         """
+        self.credentials = credentials
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/connection.py]: Done

File: src/cbc_sdk/connection.py [hunk 9]
```diff
@@ -228,6 +234,22 @@ def __init__(self,
...
+                log.debug("Fetching new bearer token as the previous token expired")
+                self.token = self.credentials.get_token()
+
+                # Update Headers
+                self.token_header.update({"Authorization": f"Bearer {self.token}"})
+                self.session.headers.update({"Authorization": f"Bearer {self.token}"})
+                r.request.headers["Authorization"] = self.session.headers["Authorization"]
+
+                # Resend Request
+                return self.session.send(r.request)
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/connection.py]: Done

**Review by rmcqueen-cb — CHANGES_REQUESTED** [review]
Minor fit and finish.

**Review by emitreva-cb — APPROVED** [review]

**Review by kebringer-cb — COMMENTED** [review]
Authentication page of read the docs needs updating

File: src/cbc_sdk/credentials.py [hunk 10]
```diff
@@ -158,6 +178,65 @@ def to_dict(self):
...
+        """
+        if self._token_type == "UNKNOWN":
+            return None
+        elif self._token_type == "API_KEY":
+            return self.get_value(CredentialValue.TOKEN)
+
+        csp_url = self.get_value(CredentialValue.CSP_URL_OVERRIDE).rstrip("/")
+
+        if self._token_type == "API_TOKEN":
+            API_TOKEN_URL = f"{csp_url}/csp/gateway/am/api/auth/api-tokens/authorize"
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/credentials.py]: Done

File: src/cbc_sdk/helpers.py [hunk 11]
```diff
@@ -53,7 +53,11 @@ def build_cli_parser(description="Cb Example Script"):
     parser = argparse.ArgumentParser(description=description)
 
     parser.add_argument("--cburl", help="CB server's URL.  e.g., http://127.0.0.1 ")
-    parser.add_argument("--apitoken", help="API Token for Carbon Black server")
+    parser.add_argument("--apitoken", help="API Token for VMware Carbon Black Cloud")
+    parser.add_argument("--csp-api-token", help="CSP API Token for VMware Carbon Black Cloud")
+    parser.add_argument("--csp-oauth-app-id", help="CSP OAuth App ID for VMware Carbon Black Cloud")
+    parser.add_argument("--csp-oauth-app-secret", help="CSP OAuth App Secret for VMware Carbon Black Cloud")
+    parser.add_argument("--csp-url-override", help="CSP URL override default value https://console.cloud.vmware.com")
```
[the remark this answers is not in the document]
Comment (avanbrunt-cb) [reply src/cbc_sdk/helpers.py]: Done

**Review by rmcqueen-cb — APPROVED** [review]

**Review by abowersox-cb — APPROVED** [review]

**Review by kebringer-cb — APPROVED** [review]

**Pull request merged**